lib/utils.ts

```diff
@@ -56,27 +56,91 @@ export const chain = (schema: z.ZodType, path: string[], register: RegisterFn):
+export const getDeepProp = <T extends Obj>(obj: T, path: string[]): unknown => {
+  // No path left, return the current value
+  if (path.length === 0 || obj === undefined || obj === null) return obj
+
+  // Get head of path
   const [head, ...tail] = path
-  const next = obj[head]
-  if (typeof next !== 'object' && next !== null)
+
+  // Is the current value an array? Can we coerce the key to a number?
+  if (Array.isArray(obj)) {
+    const key = parseInt(head)
+    if (!isNaN(key)) {
+      return getDeepProp<T>(obj[key] as T, tail)
+    } else {
+      return undefined
+    }
+  } else {
+    // Get value from obj
+    const next = obj[head]
+
+    // Is it a value we can index?
+    if (typeof next === 'object' && next !== null) {
+      return getDeepProp(next as T, tail)
+    }
+
     return next
-  return getDeepProp<T>(next as T, tail)
+  }
 }
 
-/** Mutate an object's deeply nested property */
-export const setDeepProp = <T extends Record<string, unknown>>(obj: T, path: string[], value: unknown): T => {
-  if (path.length === 0) throw new Error()
+type Obj = Record<string, unknown> | unknown[]
+
+/**
+ * Mutate an object's deeply nested property
+ * Will respect existing arrays and coerce keys to integers when applicable
+ * Cannot currently create new arrays when keys are integer. Unsure if this is wanted.
+ **/
+export const setDeepProp = <T extends Obj>(obj: T, path: string[], value: unknown): T => {
+  // Finished recursing and didn't set?
+  if (path.length === 0) throw new Error('Empty path')
 
+  // Get next part of path
   const [head, ...tail] = path
+
+  // If we are at the end of the path, set the value
   if (tail.length === 0) {
-    return {...obj, [head as keyof T]: value as T[keyof T] }
-  } else {
-    if (obj[head] === undefined) return { ...obj, [head]: setDeepProp({}, tail, value) }
-    if (typeof obj[head] !== 'object') throw new Error()
-    return { ...obj, [head]: setDeepProp(obj[head] as T, tail, value) }
+    // Is the object at this depth an array?
+    if (Array.isArray(obj)) {
+      const key = parseInt(head)
+      if (!isNaN(key)) {
+        return obj.map((v, i) => i === key ? value : v) as T
+      } else {
+        throw new Error('Object at current depth is an array but the key is not an integer')
+      }
+    } else {
+      return {...obj, [head as keyof T]: value as T[keyof T] }
+    }
+  }
+
+  // Is the object at this depth an array?
+  if (Array.isArray(obj)) {
+    // Attempt to coerce next key as a number
+    const key = parseInt(head)
+    if (!isNaN(key)) {
+      return obj.map((v, i) => i === key ? setDeepProp(obj[key as keyof T] as T, tail, value) : v) as T
+    } else {
+      throw new Error('Object at current depth is an array but the key is not an integer')
+    }
+  }
+
+  // Is there nothing at this point?
+  // Create an object for us to continue traversing into
+  if (obj[head] === undefined) {
+    // TODO: do we want this to be an array if the key is coercable as an integer?
```

Comment:
I think we do want this, as initially the data might just be an empty object. Alternatively, we might be able to restructure the `path` so it contains info about arrays, that should be relatively simple. Happy to merge this now and then work on that afterwards